### PR TITLE
Manually inspect colcon index to find CI underlay packages

### PR DESCRIPTION
**Pull request by cottsay**
Using colcon to enumerate packages in the underlay is not sufficient to detect non-ROS packages which were installed there. However, this information was captured in colcon's index when the workspace was built.

We don't need to worry about this for catkin-built workspaces because all of the packages built there were ROS packages, which catkin_pkg can detect.

Resolves ros2/ros_buildfarm_config#30

### CODE REVIEW

**Comment by cottsay** [comment]
Without this change: [Build Status](http://build.ros2.org/view/Eci/job/Eci__overlay_ubuntu_bionic_amd64/14/)
With this change: [Build Status](http://build.ros2.org/view/Eci/job/Eci__overlay_ubuntu_bionic_amd64/18/)

**Comment by nuclearsandwich** [comment]
> We don't need to worry about this for catkin-built workspaces because all of the packages built there were ROS packages, which catkin_pkg can detect.

Is this something that is currently true because we don't need to build anything but catkin packages or is catkin_make_isolated itself unable to build non-catkin CMake packages?

I'm okay if we decide there's not a way to support this with catkin_make_isolated builds but I want to clear if it's a new limitation that we're not currently hitting or an additional reason why something not currently possible will remain so.

File: scripts/ci/generate_install_lists.py [hunk 1]
```diff
@@ -84,7 +84,13 @@ def main(argv=sys.argv[1:]):
         for package_root in args.package_root[0:-1]:
             print("Crawling for packages in '%s'" % package_root)
             underlay_pkgs.update(find_packages(package_root))
-            all_underlay_pkg_names |= locate_packages(package_root).keys()
+
+            # Check for a colcon index for non-ROS package detection
+            colcon_index = os.path.join(package_root, 'colcon-core', 'packages')
```
Comment (dirk-thomas) [inline scripts/ci/generate_install_lists.py]: I am wondering why `package_root` points to the `share` directory?

Comment (cottsay) [reply scripts/ci/generate_install_lists.py]: This might be a victim of over-reusing the common argument adding methods. Here in the `generate_install_lists.py` script, `package_root` means something more like "package search path".

Here is where those search paths are computed: https://github.com/ros-infrastructure/ros_buildfarm/blob/master/ros_buildfarm/templates/ci/create_workspace.Dockerfile.em#L85

I can rename the argument to make it more clear if you like.

Comment (dirk-thomas) [reply scripts/ci/generate_install_lists.py]: No need to do it in this PR. I was just partly confused and partly curious.

**Comment by cottsay** [comment]
>> ...all of the packages built there were ROS packages...
>
> Is this something that is currently true because we don't need to build anything but catkin packages or is catkin_make_isolated itself unable to build non-catkin CMake packages?

I didn't express this correctly. As far as I know, there is no way to build a package with catkin without a `package.xml`. Assuming that `package.xml` gets installed like it should, `catkin_pkg` will always be able to detect the package in the `share` directory of the underlay space. So even for plain CMake packages, a `package.xml` is used.

tl;dr - This change is necessary since colcon can build and install packages without a `package.xml`, which catkin cannot do.
